File: AgentLens/Views/Insights/InsightsBuiltInTemplates.swift
import Foundation
import OpenBurnBarCore

/// Eight ready-to-go canvas templates the user can stamp.
enum InsightsBuiltInTemplates {

    static var all: [InsightCanvasTemplate] {
        [today, costAudit, agentFocus, modelFocus, useCaseLibrary, quotaHealth, quarterlyReview, anomalies]
    }

    // MARK: - Today

    static var today: InsightCanvasTemplate {
        .init(
            id: "today",
            title: "Today",
            summary: "A daily snapshot of cost, sessions, cache, and your top model.",
            symbolName: "sun.max.fill",
            theme: .aurora,
            widgets: [
                widget(.kpiTile, title: "Cost", binding: .kpi(metric: .totalCost, window: .today)),
                widget(.kpiTile, title: "Sessions", binding: .kpi(metric: .totalSessions, window: .today)),
                widget(.kpiTile, title: "Cache hit", binding: .kpi(metric: .cacheHitRate, window: .today)),
                widget(.kpiTile, title: "Tokens", binding: .kpi(metric: .totalTokens, window: .today)),
                widget(.timeSeriesLine, title: "Today by provider",
                       binding: .timeSeries(metric: .cost, dimension: .provider, window: .today),
                       spec: .timeSeries(.init(style: .line))),
                widget(.heatmap, title: "When you worked",
                       binding: .heatmap(metric: .sessions, window: .today)),
                widget(.narrative, title: "Today's narrative",
                       binding: .narrative(.init(headline: "Today",
                                                  body: "A daily summary will appear here once you run an investigation.")),
                       spec: .narrative(.init())),
                widget(.quotaPulse, title: "Quota pulse",
                       binding: .quota(providerKey: nil))
            ],
            layout: defaultLayout(),
            filter: InsightFilter(window: .today)
        )
    }

    // MARK: - Cost audit

    static var costAudit: InsightCanvasTemplate {
        .init(
            id: "cost-audit-7d",
            title: "Cost Audit (7d)",
            summary: "Where the money went last week.",
            symbolName: "dollarsign.circle.fill",
            theme: .ember,
            widgets: [
                widget(.kpiTile, title: "7d cost", binding: .kpi(metric: .totalCost, window: .last7d)),
                widget(.kpiTile, title: "Avg / session",
                       binding: .kpi(metric: .avgCostPerSession, window: .last7d)),
                widget(.timeSeriesLine, title: "Cost trend",
                       binding: .timeSeries(metric: .cost, dimension: .provider, window: .last7d),
                       spec: .timeSeries(.init(style: .area))),
                widget(.treemap, title: "Spend by provider × model",
                       binding: .distribution(metric: .cost, dimension: .model, window: .last7d),
                       spec: .distribution(.init(style: .treemap))),
                widget(.scatter, title: "Efficiency frontier",
                       binding: .scatter(xMetric: .tokens, yMetric: .cost, dimension: .model, window: .last7d)),
                widget(.barRanking, title: "Top spenders",
                       binding: .ranking(metric: .cost, dimension: .model, limit: 10, window: .last7d)),
                widget(.forecast, title: "Next 7d projection",
                       binding: .forecast(metric: .cost, horizonDays: 7)),
                widget(.recommendation, title: "Top recommendation",
                       binding: .recommendation(.init(headline: "Run an investigation",
                                                       rationale: "Recommendations populate after the first investigation.",
                                                       action: "Ask the composer 'how do I spend less?'")))
            ],
            layout: defaultLayout(),
            filter: InsightFilter(window: .last7d)
        )
    }

    // MARK: - Agent focus

    static var agentFocus: InsightCanvasTemplate {
        .init(
            id: "agent-focus",
            title: "Agent Focus",
            summary: "What each agent is being used for.",
            symbolName: "person.crop.square.filled.and.at.rectangle",
            theme: .whimsy,
            widgets: [
                widget(.agentFocusMatrix, title: "Focuses by agent",
                       binding: .agentFocusMatrix(window: .last30d)),
                widget(.radar, title: "Top agents — capability fingerprint",
                       binding: .radar(target: .allAgents, window: .last30d)),
                widget(.useCaseCluster, title: "Common use cases",
                       binding: .useCaseClusters(window: .last30d)),
                widget(.drilldownList, title: "Recent sessions",
                       binding: .drilldown(limit: 20))
            ],
            layout: defaultLayout(),
            filter: InsightFilter(window: .last30d)
        )
    }

    // MARK: - Model focus

    static var modelFocus: InsightCanvasTemplate {
        .init(
            id: "model-focus",
            title: "Model Focus",
            summary: "Where each model excels.",
            symbolName: "cpu.fill",
            theme: .mercury,
            widgets: [
                widget(.donut, title: "Model mix",
                       binding: .distribution(metric: .tokens, dimension: .model, window: .last30d)),
                widget(.modelFocusMatrix, title: "Focuses by model",
                       binding: .modelFocusMatrix(window: .last30d)),
                widget(.scatter, title: "Cost-per-Mtoken vs. volume",
                       binding: .scatter(xMetric: .tokens, yMetric: .costPerMtoken,
                                          dimension: .model, window: .last30d)),
                widget(.narrative, title: "Model shift",
                       binding: .narrative(.init(headline: "—", body: "Run an investigation for narrative.")),
                       spec: .narrative(.init()))
            ],
            layout: defaultLayout(),
            filter: InsightFilter(window: .last30d)
        )
    }

    // MARK: - Use-case library

    static var useCaseLibrary: InsightCanvasTemplate {
        .init(
            id: "use-case-library",
            title: "Use-Case Library",
            summary: "Tags, clusters, and examples.",
            symbolName: "tag.circle.fill",
            theme: .aurora,
            widgets: [
                widget(.useCaseCluster, title: "Use case clusters",
                       binding: .useCaseClusters(window: .last90d)),
                widget(.agentFocusMatrix, title: "Agent × focus",
                       binding: .agentFocusMatrix(window: .last90d)),
                widget(.drilldownList, title: "Top sessions",
                       binding: .drilldown(limit: 20))
            ],
            layout: defaultLayout(),
            filter: InsightFilter(window: .last90d)
        )
    }

    // MARK: - Quota health

    static var quotaHealth: InsightCanvasTemplate {
        .init(
            id: "quota-health",
            title: "Quota Health",
            summary: "How close you are to your provider caps.",
            symbolName: "gauge.with.dots.needle.67percent",
            theme: .ember,
            widgets: [
                widget(.quotaPulse, title: "Quota pulse", binding: .quota(providerKey: nil)),
                widget(.recommendation, title: "Headroom suggestion",
                       binding: .recommendation(.init(headline: "—", rationale: "—", action: "Ask the composer."))),
                widget(.timeSeriesLine, title: "Usage trend",
                       binding: .timeSeries(metric: .tokens, dimension: .provider, window: .last7d))
            ],
            layout: defaultLayout(),
            filter: InsightFilter(window: .last7d)
        )
    }

    // MARK: - Quarterly review

    static var quarterlyReview: InsightCanvasTemplate {
        .init(
            id: "quarterly-review",
            title: "Quarterly Review",
            summary: "90 days at a glance.",
            symbolName: "calendar",
            theme: .mercury,
            widgets: [
                widget(.kpiTile, title: "90d cost", binding: .kpi(metric: .totalCost, window: .last90d)),
                widget(.kpiTile, title: "Sessions", binding: .kpi(metric: .totalSessions, window: .last90d)),
                widget(.timeSeriesLine, title: "Cost over 90d",
                       binding: .timeSeries(metric: .cost, dimension: .provider, window: .last90d),
                       spec: .timeSeries(.init(style: .area))),
                widget(.cohort, title: "Cohort retention",
                       binding: .cohort(window: .last90d)),
                widget(.barRanking, title: "Top 10 models",
                       binding: .ranking(metric: .cost, dimension: .model, limit: 10, window: .last90d)),
                widget(.narrative, title: "Highlights",
                       binding: .narrative(.init(headline: "—", body: "Run an investigation.")),
                       spec: .narrative(.init()))
            ],
            layout: defaultLayout(),
            filter: InsightFilter(window: .last90d)
        )
    }

    // MARK: - Anomalies

    static var anomalies: InsightCanvasTemplate {
        .init(
            id: "anomalies",
            title: "Anomalies",
            summary: "Outlier days, spikes, and dips.",
            symbolName: "exclamationmark.triangle.fill",
            theme: .ember,
            widgets: [
                widget(.anomalyTable, title: "Anomaly table",
                       binding: .anomaly(window: .last90d)),
                widget(.timeSeriesLine, title: "Cost with anomalies",
                       binding: .timeSeries(metric: .cost, dimension: nil, window: .last90d)),
                widget(.drilldownList, title: "Sessions on outlier days",
                       binding: .drilldown(limit: 12)),
                widget(.narrative, title: "Per-anomaly explanation",
                       binding: .narrative(.init(headline: "—", body: "—")),
                       spec: .narrative(.init()))
            ],
            layout: defaultLayout(),
            filter: InsightFilter(window: .last90d)
        )
    }

    // MARK: - Helpers

    private static func widget(_ kind: InsightWidgetKind,
                               title: String,
                               binding: InsightDataBinding,
                               spec: InsightWidgetSpec? = nil) -> InsightWidget {
        InsightWidget(
            kind: kind,
            title: title,
            spec: spec ?? defaultSpec(for: kind),
            dataBinding: binding
        )
    }

    private static func defaultSpec(for kind: InsightWidgetKind) -> InsightWidgetSpec {
        switch kind {
        case .kpiTile: return .kpiTile(.init(metricLabel: ""))
        case .timeSeriesLine: return .timeSeries(.init(style: .line))
        case .timeSeriesArea: return .timeSeries(.init(style: .area))
        case .streamGraph: return .timeSeries(.init(style: .stream))
        case .barRanking: return .ranking(.init())
        case .donut: return .distribution(.init(style: .donut))
        case .treemap: return .distribution(.init(style: .treemap))
        case .heatmap: return .heatmap(.init())
        case .scatter: return .scatter(.init())
        case .sankey: return .sankey(.init())
        case .radar: return .radar(.init())
        case .cohort: return .cohort(.init())
        case .funnel: return .funnel(.init())
        case .quotaPulse: return .quotaPulse(.init())
        case .forecast: return .forecast(.init())
        case .anomalyTable: return .anomalyTable(.init())
        case .narrative: return .narrative(.init())
        case .recommendation: return .recommendation(.init())
        case .useCaseCluster: return .useCaseCluster(.init())
        case .agentFocusMatrix: return .agentFocusMatrix(.init())
        case .modelFocusMatrix: return .modelFocusMatrix(.init())
        case .drilldownList: return .drilldownList(.init())
        case .mermaid: return .mermaid(.init())
        case .ascii: return .ascii(.init())
        case .composed: return .composed(.init(children: []))
        case .error: return .error(.init(message: ""))
        }
    }

    /// Layout that places widgets row-major into 12 columns.
    private static func defaultLayout() -> InsightLayout {
        InsightLayout(columnCount: 12, rowHeight: 96, gap: 12)
    }
}

File: windows/app/OpenBurnBar.App.Presentation/Insights/InsightsBuiltInTemplates.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace OpenBurnBar.App.Presentation.Insights;

/// <summary>
/// The eight ready-to-go canvas templates the user can stamp — a port of the macOS
/// <c>InsightsBuiltInTemplates</c> (<c>AgentLens/Views/Insights/InsightsBuiltInTemplates.swift</c>).
///
/// Production default: missing live data resolves to honest empty widgets
/// (<see cref="InsightEmptyData"/>). Deterministic <see cref="InsightSampleData"/> is
/// only injected when <see cref="SampleFallbackEnabled"/> is explicitly on (opt-in demo
/// via <c>OPENBURNBAR_SAMPLE_MODE</c>, or unit tests that exercise sample generators).
/// Icon glyphs are Segoe MDL2 Assets code points (approximate; final icon parity is a
/// design pass).
/// </summary>
public static class InsightsBuiltInTemplates
{
    private static readonly object Gate = new();
    private static Func<InsightWidgetKind, int, InsightWidgetData?>? _realDataResolver;
    private static bool _sampleFallbackEnabled;
    private static IReadOnlyList<InsightCanvasTemplate>? _all;
    private static int _resolverVersion;

    /// <summary>
    /// Optional override for real data resolution. When set, the resolver result wins
    /// over empty/sample fallbacks. The app-level page installs a production resolver
    /// that reads KPI tiles from the SQLCipher / cloud usage summary.
    /// </summary>
    public static Func<InsightWidgetKind, int, InsightWidgetData?>? RealDataResolver
    {
        get
        {
            lock (Gate)
            {
                return _realDataResolver;
            }
        }

        set
        {
            lock (Gate)
            {
                _realDataResolver = value;
                _all = null;
                _resolverVersion++;
            }
        }
    }

    /// <summary>
    /// When true, unresolved widgets receive deterministic <see cref="InsightSampleData"/>.
    /// Defaults to <c>false</c> so production paths cannot silently paint demo series.
    /// The WinUI Insights page sets this from <c>RuntimeDataMode.SampleModeEnabled</c>;
    /// unit tests that need sample generators enable it in a try/finally.
    /// </summary>
    public static bool SampleFallbackEnabled
    {
        get
        {
            lock (Gate)
            {
                return _sampleFallbackEnabled;
            }
        }

        set
        {
            lock (Gate)
            {
                if (_sampleFallbackEnabled == value)
                {
                    return;
                }

                _sampleFallbackEnabled = value;
                _all = null;
                _resolverVersion++;
            }
        }
    }

    /// <summary>All templates, in gallery order (matches the macOS ordering).</summary>
    public static IReadOnlyList<InsightCanvasTemplate> All
    {
        get
        {
            Func<InsightWidgetKind, int, InsightWidgetData?>? resolver;
            bool sampleFallback;
            int version;
            lock (Gate)
            {
                if (_all is { } cached)
                {
                    return cached;
                }

                resolver = _realDataResolver;
                sampleFallback = _sampleFallbackEnabled;
                version = _resolverVersion;
            }

            IReadOnlyList<InsightCanvasTemplate> built = BuildAll(resolver, sampleFallback);

            lock (Gate)
            {
                if (_all is { } cached)
                {
                    return cached;
                }

                if (_resolverVersion == version)
                {
                    _all = built;
                    return _all;
                }
            }

            return All;
        }
    }

    /// <summary>Resolve a template by its stable string id.</summary>
    public static InsightCanvasTemplate? Find(string? id)
        => id is null ? null : All.FirstOrDefault(t => t.Id == id);

    private static IReadOnlyList<InsightCanvasTemplate> BuildAll(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback) => new[]
    {
        Today(resolver, sampleFallback),
        CostAudit(resolver, sampleFallback),
        AgentFocus(resolver, sampleFallback),
        ModelFocus(resolver, sampleFallback),
        UseCaseLibrary(resolver, sampleFallback),
        QuotaHealth(resolver, sampleFallback),
        QuarterlyReview(resolver, sampleFallback),
        Anomalies(resolver, sampleFallback),
    };

    private static InsightWidget W(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback,
        InsightWidgetKind kind,
        string title,
        int seed)
    {
        InsightWidgetData? realData = resolver?.Invoke(kind, seed);
        if (realData is not null)
        {
            return InsightWidget.Create(kind, title, realData);
        }

        InsightWidgetData data = sampleFallback
            ? InsightSampleData.ForKind(kind, seed)
            : InsightEmptyData.ForKind(kind, seed);
        return InsightWidget.Create(kind, title, data);
    }

    private static InsightCanvasTemplate Today(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback) => new(
        Id: "today",
        Title: "Today",
        Summary: "A daily snapshot of cost, sessions, cache, and your top model.",
        Glyph: "\uE706", // brightness / sun
        Theme: InsightTheme.Aurora,
        Widgets: new List<InsightWidget>
        {
            W(resolver, sampleFallback, InsightWidgetKind.KpiTile, "Cost", 1),
            W(resolver, sampleFallback, InsightWidgetKind.KpiTile, "Sessions", 2),
            W(resolver, sampleFallback, InsightWidgetKind.KpiTile, "Cache hit", 3),
            W(resolver, sampleFallback, InsightWidgetKind.KpiTile, "Tokens", 4),
            W(resolver, sampleFallback, InsightWidgetKind.TimeSeriesLine, "Today by provider", 5),
            W(resolver, sampleFallback, InsightWidgetKind.Heatmap, "When you worked", 6),
            W(resolver, sampleFallback, InsightWidgetKind.Narrative, "Today's narrative", 7),
            W(resolver, sampleFallback, InsightWidgetKind.QuotaPulse, "Quota pulse", 8),
        });

    private static InsightCanvasTemplate CostAudit(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback) => new(
        Id: "cost-audit-7d",
        Title: "Cost Audit (7d)",
        Summary: "Where the money went last week.",
        Glyph: "\uE8C7", // money
        Theme: InsightTheme.Ember,
        Widgets: new List<InsightWidget>
        {
            W(resolver, sampleFallback, InsightWidgetKind.KpiTile, "7d cost", 11),
            W(resolver, sampleFallback, InsightWidgetKind.KpiTile, "Avg / session", 12),
            W(resolver, sampleFallback, InsightWidgetKind.TimeSeriesArea, "Cost trend", 13),
            W(resolver, sampleFallback, InsightWidgetKind.Donut, "Spend by model", 14),
            W(resolver, sampleFallback, InsightWidgetKind.Scatter, "Efficiency frontier", 15),
            W(resolver, sampleFallback, InsightWidgetKind.BarRanking, "Top spenders", 16),
            W(resolver, sampleFallback, InsightWidgetKind.Forecast, "Next 7d projection", 17),
            W(resolver, sampleFallback, InsightWidgetKind.Recommendation, "Top recommendation", 18),
        });

    private static InsightCanvasTemplate AgentFocus(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback) => new(
        Id: "agent-focus",
        Title: "Agent Focus",
        Summary: "What each agent is being used for.",
        Glyph: "\uE716", // people
        Theme: InsightTheme.Whimsy,
        Widgets: new List<InsightWidget>
        {
            W(resolver, sampleFallback, InsightWidgetKind.Heatmap, "Focuses by agent", 21),
            W(resolver, sampleFallback, InsightWidgetKind.Radar, "Top agents — capability fingerprint", 22),
            W(resolver, sampleFallback, InsightWidgetKind.Donut, "Common use cases", 23),
            W(resolver, sampleFallback, InsightWidgetKind.BarRanking, "Recent sessions", 24),
        });

    private static InsightCanvasTemplate ModelFocus(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback) => new(
        Id: "model-focus",
        Title: "Model Focus",
        Summary: "Where each model excels.",
        Glyph: "\uE964", // cpu / chip
        Theme: InsightTheme.Mercury,
        Widgets: new List<InsightWidget>
        {
            W(resolver, sampleFallback, InsightWidgetKind.Donut, "Model mix", 31),
            W(resolver, sampleFallback, InsightWidgetKind.Heatmap, "Focuses by model", 32),
            W(resolver, sampleFallback, InsightWidgetKind.Scatter, "Cost-per-Mtoken vs. volume", 33),
            W(resolver, sampleFallback, InsightWidgetKind.Narrative, "Model shift", 34),
        });

    private static InsightCanvasTemplate UseCaseLibrary(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback) => new(
        Id: "use-case-library",
        Title: "Use-Case Library",
        Summary: "Tags, clusters, and examples.",
        Glyph: "\uE8EC", // tag
        Theme: InsightTheme.Aurora,
        Widgets: new List<InsightWidget>
        {
            W(resolver, sampleFallback, InsightWidgetKind.Donut, "Use case clusters", 41),
            W(resolver, sampleFallback, InsightWidgetKind.Heatmap, "Agent × focus", 42),
            W(resolver, sampleFallback, InsightWidgetKind.BarRanking, "Top sessions", 43),
        });

    private static InsightCanvasTemplate QuotaHealth(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback) => new(
        Id: "quota-health",
        Title: "Quota Health",
        Summary: "How close you are to your provider caps.",
        Glyph: "\uE9D9", // gauge
        Theme: InsightTheme.Ember,
        Widgets: new List<InsightWidget>
        {
            W(resolver, sampleFallback, InsightWidgetKind.QuotaPulse, "Quota pulse", 51),
            W(resolver, sampleFallback, InsightWidgetKind.Recommendation, "Headroom suggestion", 52),
            W(resolver, sampleFallback, InsightWidgetKind.TimeSeriesLine, "Usage trend", 53),
        });

    private static InsightCanvasTemplate QuarterlyReview(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback) => new(
        Id: "quarterly-review",
        Title: "Quarterly Review",
        Summary: "90 days at a glance.",
        Glyph: "\uE787", // calendar
        Theme: InsightTheme.Mercury,
        Widgets: new List<InsightWidget>
        {
            W(resolver, sampleFallback, InsightWidgetKind.KpiTile, "90d cost", 61),
            W(resolver, sampleFallback, InsightWidgetKind.KpiTile, "Sessions", 62),
            W(resolver, sampleFallback, InsightWidgetKind.TimeSeriesArea, "Cost over 90d", 63),
            W(resolver, sampleFallback, InsightWidgetKind.Funnel, "Conversion", 64),
            W(resolver, sampleFallback, InsightWidgetKind.BarRanking, "Top 10 models", 65),
            W(resolver, sampleFallback, InsightWidgetKind.Narrative, "Highlights", 66),
        });

    private static InsightCanvasTemplate Anomalies(
        Func<InsightWidgetKind, int, InsightWidgetData?>? resolver,
        bool sampleFallback) => new(
        Id: "anomalies",
        Title: "Anomalies",
        Summary: "Outlier days, spikes, and dips.",
        Glyph: "\uE7BA", // warning
        Theme: InsightTheme.Ember,
        Widgets: new List<InsightWidget>
        {
            W(resolver, sampleFallback, InsightWidgetKind.BarRanking, "Anomaly table", 71),
            W(resolver, sampleFallback, InsightWidgetKind.TimeSeriesLine, "Cost with anomalies", 72),
            W(resolver, sampleFallback, InsightWidgetKind.Sankey, "Sessions on outlier days", 73),
            W(resolver, sampleFallback, InsightWidgetKind.Narrative, "Per-anomaly explanation", 74),
        });
}
